docs/proposals/20240807-in-place-updates.md

```diff
@@ -0,0 +1,488 @@
+5. Machine Controller: make machine is updated.
+
+The following sections dive deep into these steps, zooming in into the different component interactions and defining how the main error cases are handled.
+
+### Defining the Update Plan
+
+```mermaid
+sequenceDiagram
+    participant Operator
+
+    box Management Cluster
+        participant apiserver as kube-api server
+        participant capi as CP/MD Controller
+        participant hook as External updater 1
+        participant hook2 as External updater 2
+    end
+
+    Operator->>+apiserver: make changes to CP/MD <br>(ExternalUpdate strategy)
+    apiserver->>+capi: Notify changes
+    apiserver->>-Operator: OK
+    capi->>+hook: Can update?
+    hook->>capi: Set of changes
+    capi->>+hook2: Can update?
+    hook2->>capi: Set of changes
+    alt all changes covered?
+        capi->>apiserver: Define Update Plan
+    else
+        alt fallback strategy?
+            capi->>apiserver: Re-create machines
+        else
+            capi->>apiserver: Marked update as failed
+        end
+    end
+```
+
+Both `KCP` and `MachineDeployment` controllers follow a similar pattern around updates, they first detect if an update is required and then based on the configured strategy follow the appropiate update logic (note that today there is only one valid strategy, `RollingUpdate`).
+
+With `ExternalUpdate` strategy, CAPI controllers will compute the set of desired changes and iterate over the registered external updaters, requesting through the Runtime Hook the set of changes each updater can handle. The changes supported by an updater can be the complete set of desired changes, a subset of them or an empty set, signaling it cannot handle any of the desired changes.
+
+On each iteration, the controller will substract from the set of desired changes the subset returned by the updater. 
+
+If this set is reduced to zero, then CAPI will determine that the update can be performed using the external strategy. CAPI will define the update plan as a list of sequential external updaters in a particular order and proceed to execute it. The update plan will be stored in the Machine object as an array of strings (the names of the selected external updaters).
+
+If after iterating over all external updaters the remaining set still contains uncovered changes, CAPI will determine the desired state cannot be reached through external updaters. If a fallback rolling update strategy has been configured (this is optional), CAPI will replace the machines. If no fallback strategy is configured, we will surface the issue in the resource status. Machines will remain unchanged and the desired state won't be reached unless remediated by the user. Depending on the scenario, users can: ammend the desired state to something that the registered updaters can cover, register additional updaters capable of handling the desired changes or simply enable the fallback strategy.
```

Comment:
How is the order of external upgraders can be defined? Since there will be implicit requirements which will make them dependent on each other.
 
Since the idea is to iterate over an array of upgraders, this should have support for multiple iterations, and more clever mechanism than substraction. One iteration will not be enough to mark desired state unreachable.

Reply:
With the current proposal, updaters need to be independent in order to be scheduled in the same upgrade plan. Updaters just look at the set of required changes and tell capi what is the subset of changes they can take care of. And they need to be capable of updating those fields regardless of how many other updaters are scheduled and no matter if they run before or after.

If for some reason an updater needs certain fields to be updated first before being able to execute its update, then two update plans will be needed, hence the change would need to be performed by the user in two phases.

We could (probably in future iterations) add a "priority" property to the updaters that would help order updaters when they have overlapping functions. However this would be a global priority and not relative between updaters.

Now all that said, this is a hat we are proposing, which might not cover all usecases. Do you have a particular usecase where order matters and updaters **must** be dependent on each other.